forest/cli/src/cli/wallet_cmd.rs

```diff
@@ -241,3 +276,246 @@ impl WalletCommands {
         }
     }
 }
+
+fn format_balance_string(num: Decimal, mode: FormattingMode) -> String {
+    let units = [
+        "atto", "femto", "pico", "nano", "micro", "milli", "", "kilo", "mega", "giga", "tera",
+        "peta", "exa", "zetta", "yotta",
+    ];
+    let orig = num;
+
+    let mut num = num;
+    let mut unit_index = 0;
+    while num >= dec!(1000.0) && unit_index < units.len() - 1 {
+        num /= dec!(1000.0);
+        unit_index += 1;
+    }
+
+    match mode {
+        FormattingMode::ExactFixed => {
+            let fil = orig / dec!(1e18);
+            format!("{fil} FIL")
+        }
+        FormattingMode::NotExactFixed => {
+            let fil_orig = orig / dec!(1e18);
+            let fil = fil_orig.round_dp_with_strategy(3, RoundingStrategy::MidpointAwayFromZero);
+            let mut res = format!("{} FIL", fil.trunc());
+            if fil != fil_orig {
+                res.insert(0, '~');
+            }
+            res
+        }
+        FormattingMode::ExactNotFixed => format!("{num:0} {} FIL", units[unit_index]),
+        FormattingMode::NotExactNotFixed => {
+            let mut fil = num.round_dp_with_strategy(3, RoundingStrategy::MidpointAwayFromZero);
+            if fil == fil.trunc() {
+                fil = fil.trunc();
+            }
+            let mut res = format!("{} {} FIL", fil, units[unit_index]);
+
+            if fil != num {
+                res.insert(0, '~');
+            }
+
+            res
+        }
+    }
+}
+
+fn bool_pair_to_mode(exact: bool, fixed: bool) -> FormattingMode {
+    if exact && fixed {
+        FormattingMode::ExactFixed
+    } else if !exact && fixed {
+        FormattingMode::NotExactFixed
+    } else if exact && !fixed {
+        FormattingMode::ExactNotFixed
+    } else {
+        FormattingMode::NotExactNotFixed
+    }
+}
+
+#[test]
+fn exact_balance_fixed_unit() {
+    let mut token_string = TokenAmount::from_atto(100).atto().to_string();
+    assert_eq!(
+        format_balance_string(
+            token_string.parse::<Decimal>().unwrap(),
+            bool_pair_to_mode(true, true)
+        ),
+        "0.0000000000000001 FIL"
+    );
+
+    token_string = TokenAmount::from_atto(12465).atto().to_string();
+    assert_eq!(
+        format_balance_string(
+            token_string.parse::<Decimal>().unwrap(),
+            bool_pair_to_mode(true, true)
+        ),
+        "0.000000000000012465 FIL"
+    );
+}
+
+#[test]
+fn not_exact_balance_fixed_unit() {
+    let mut token_string = TokenAmount::from_atto(100).atto().to_string();
```

Comment:
That's a lot of boilerplate code. We could shorten it significantly with this pattern https://github.com/ChainSafe/forest/blob/main/forest/cli/src/cli/mod.rs#L194

Reply:
Also, there's no point in creating a `token_string`. We can use the decimal directly; I assume `BigInt` is properly tested on its own.

Reply:
It's done not to test `BigInt`, but to recreate real life usage.

Reply:
What real life usage? We are not passing those `BigInt`s anywhere. It's not increasing code coverage at all.

The pattern I had in mind was
```
let cases = [
(input, expected_output),
(input, expected_output),
];

assert_eq!(expected_output, tested_method(input);
```
It would shorten the code significantly. There's no point in calling `parse` or `from` everywhere.

Reply:
I applied the pattern